### PR TITLE
fix(metrics): Fix BinaryAccuracy metric to handle boolean inputs

### LINKED ISSUE #20178: Wrong binary accuracy with Jax
I have some very strange results out of the `

Consider the code below:
```
import os
os.environ["KERAS_BACKEND"] = "jax"
import keras


inp = keras.Input(shape=(1,))
out = inp > 0.5
mm = keras.Model(inputs=inp, outputs=out) 

x = np.random.rand(32, 1)

res = mm.predict(x)
met = keras.metrics.BinaryAccuracy()
met.update_state(x>0.5, res>0.5)
met.result()
```

I would expect to get 1 every single run. Instead I get some random result (close to 0.5). 

Packages' versions (tf, keras, jax, np)
```
'2.17.0', '3.5.0', '0.4.26', '1.26.4'
```

### PR DESCRIPTION
**Pull request by harshaljanjani**
Previously, `BinaryAccuracy` would return incorrect results when given boolean inputs in JAX backend, and would raise errors in TensorFlow backend. This was because the metric expects numerical values (floats/integers) but wasn't properly handling boolean array inputs.
Fix by casting `y_true` and `y_pred` to `floatx()` in `MeanMetricWrapper.update_state()`. This ensures consistent behavior across backends and proper handling of boolean inputs.
Fixes #20178 raised in August 2024 but still persists in this version.

### CODE REVIEW

File: keras/src/metrics/reduction_metrics.py [hunk 1]
```diff
@@ -199,6 +199,9 @@ def __init__(self, fn, name=None, dtype=None, **kwargs):
             self._direction = "down"
 
     def update_state(self, y_true, y_pred, sample_weight=None):
+        y_true = backend.cast(y_true, backend.floatx())
```
Comment (fchollet) [inline keras/src/metrics/reduction_metrics.py:202]: I think we should prefer using `self.dtype`?

Comment (harshaljanjani) [reply keras/src/metrics/reduction_metrics.py:202]: Agreed, will make the change. Thanks for pointing this out!

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
